### PR TITLE
Add support for XChaCha20.

### PR DESCRIPTION
**Pull request by Yawning**
XChaCha20 is the extended 192 bit nonce variant of the ChaCha20 stream
cipher.  Like the current existing XSalsa20 implementation, only 256
bit keys are supported.

Note: As there are no existing test vectors for this algorithm, so a
simple test was generated using Andrew Moon's chacha-opt library, using
the same key and nonce as the XSalsa20 test case.

### CODE REVIEW

**Comment by DaGenix** [comment]
Thanks for submitting this! I don't have time right now to take a deeper look, but, I'll try to look at this more tomorrow.

Thanks!

**Comment by DaGenix** [comment]
Is there a particular spec that you implemented? I'm having a tricky time finding a docment describing XChaCha20.

**Comment by Yawning** [comment]
@DaGenix 

It works exactly like XSalsa20 does, except with a different underlying stream cipher.  The cascaded setup provided in [Extending the Salsa20 nonce](http://cr.yp.to/snuffle/xsalsa-20110204.pdf) is generic (as is the security proof).

The HChaCha code looks different from HSalsa (including where the subkey is derived) because the matrix was changed between Salsa20 and ChaCha20.  However the subkey offsets still belong to the positions corresponding to the constant and the input, so in terms of functionality, it's cosmetic and as far as i can tell the XSalsa/HSalsa security proof applies for this construct.

https://www.ietf.org/mail-archive/web/cfrg/current/msg04310.html does an ok job of explaining what was implemented (though I didn't bother supporting non-24 byte nonces via zero padding), and https://github.com/floodyberry/chacha-opt is the implementation I tested against.

There's a few minor cleanups I can make to the code that would improve readability, so I shall go do that.

**Comment by DaGenix** [comment]
The general policy of rust-crypto is not to take expirimental algorithms that aren't being used somewhere in the wild or don't have a formal spec. That policy isn't written down, though, so "policy" might not be the best word for this. This is definitally something I need to work on and establish a consensus with developers who contribute to the project.

However, I'm inclined to bend that "policy" a bit and merge this since its fairly self contained and since there are already algorithms in the library that don't have firm specs behind them. I would like to introduce some sort of "feature gate" type system so that users have more control over what they are compiling and also have to explicitly opt-in to expirimental algorithms. So, that would probably put XChaCha20 into the expirimental category until there is a spec by IETF or some other standards body or until its being widely used by something (if it is in wide use by something, please let me know, but my quick googling didn't turn up anything).

Anyway, long story short - I want to take one more look at this tomorow and then I intend to merge.

Thanks!

**Comment by Yawning** [comment]
@DaGenix Understandable, and I agree that such a policy makes a lot of sense.  I'm kind of sad to this day that the NaCl `box` constructs are defined in terms of XSalsa20 when XChaCha20 is objectively a better algorithm, but I assume there are historical reasons behind that (and is a rant more suited for a different location).

**Comment by DaGenix** [comment]
Thanks!
